File: Java/KadaneAlgorithm.java
import java.util.Scanner;

public class KadaneAlgorithm
{
	static int maxSubArraySum(int a[])
	{
		int size = a.length;
		int max_so_far = Integer.MIN_VALUE, max_ending_here = 0;

		for (int i = 0; i < size; i++)
		{
			max_ending_here = max_ending_here + a[i];
			if (max_so_far < max_ending_here)
				max_so_far = max_ending_here;
			if (max_ending_here < 0)
				max_ending_here = 0;
		}
		return max_so_far;
	}
	
	public static void main (String[] args)
	{
		Scanner sc = new Scanner(System.in);
		
		//Input the size of the array
        System.out.println("Enter the size of array: ");
        int n = sc.nextInt();
        
        //Initialize the array with given size
        int arr[] = new int[n];
        
        //Input the array elements
        System.out.println("Enter array elements");
        for(int i = 0; i< n; i++)
            arr[i] = sc.nextInt();
        
        sc.close();
        
		System.out.println("Maximum contiguous sum is " +
									maxSubArraySum(arr));
	}
}
/*
 * Java program for Implementation of Kadane's Algorithm 
 * Input: Size of array,array of integers 
 * Output: Sorted array in ascending order
 * ---------------------------------------- 
 * Sample Input1: 
 * 4
 * 25 9 -49 16
 * Output:
 * Maximum contiguous sum is 34
 *
 * Sample Input2:
 * 5
 * 5 4 -1 7 8
 * Output:
 * Maximum contiguous sum is 23
 *
 * Sample Input3:
 * 9
 * -2 1 -3 4 -1 2 1 -5 4
 * Output:
 * Maximum contiguous sum is 6
 * 
 * ---------------------------------------- 
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

File: C/Kadane_algorithm.c
#include <stdio.h>

// Problem based on it : https://practice.geeksforgeeks.org/problems/kadanes-algorithm-1587115620/1

int kadane_algorithm(int arr[], int n)
{
    int max_sum = 0, ans = -1e9, i; 
    // we want to choose a subarray with the maximum sum (kadane_algorithm), so its intutitive that we should try to avoid negative number
    // in case where there are all negative numbers, then we choose the maximum element
    // so maximum_sum starts from 0, ans starting from -1e9 as there is possibility that all elements of array are negative, not any negative value;
    
    // Now its very intutitive that we have to take positive numbers to maximise the sum 
    // But wait can we take negative numbers ???
    // We can take negative numbers till then we current_sum + negative_number_current >= 0
    // But why this condition only ? 
    
    // Suppose array is like 5 -2 3 
    // 5 ---> 3 (negative taken) or 0 (negative not taken, break subarray, so maximum_sum starts from 0) ----> 6 or 3
    
    // Suppose array is like 5 -6 3
    // 5 ---> -1 (negative taken) or 0 (negative not taken, break subarray, so maximum_sum starts from 0) -----> 2 or 3
    
    // I hope it clears your doubt why and when we take negative numbers in kadane_algorithm
    
    // Once again, we take negative numbers when (current_sum + negative_number_current >= 0)
    
    for(i=0; i<n; i++)
    {
        max_sum += arr[i];
            
        if(max_sum >= ans)
            ans = max_sum;
            
        if(max_sum <= 0)
            max_sum = 0;
    }
    
    return ans;
}

int main()
{
    int i, n;
    printf("Enter the number of element : \n");
    scanf("%d ", &n);
    
    int arr[n];
    
    printf("Enter the elements of the array : \n");
    
    for(i=0; i<n; i++)
    scanf("%d ", &arr[i]);
    
    printf("Elements of the array : \n");
    
    for(i=0; i<n; i++)
    printf("%d ", arr[i]);
    
    printf("\nMax subarrary sum : %d\n", kadane_algorithm(arr, n));
}

/* Testcase :

Input : 
8
-2 -3 4 -1 -2 1 5 -3

Output : 
7 

Time Complexity : 
O(N) ---> As we are traversing the whole array one time 

Space Complexity : 
O(N) ---> As we are storing the data in the array */